flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/UnboundedEventTimeOverProcessFunction.scala

```diff
@@ -0,0 +1,181 @@
+        getRuntimeContext.getExecutionConfig).asInstanceOf[TypeSerializer[Tuple2[Long, Row]]]
+    val tupleStateDescriptor: ListStateDescriptor[Tuple2[Long, Row]] =
+      new ListStateDescriptor[Tuple2[Long, Row]]("rowliststate", tupleSerializer)
+    rowState = getRuntimeContext.getListState[Tuple2[Long, Row]](tupleStateDescriptor)
+
+  }
+
+  /**
+    * Process one element from the input stream, not emit the output
+    *
+    * @param value The input value.
+    * @param ctx   The ctx to register timer or get current time
+    * @param out   The collector for returning result values.
+    *
+    */
+  override def processElement(
+     input: Row,
+     ctx:  ProcessFunction[Row, Row]#Context,
+     out: Collector[Row]): Unit = {
+
+    // discard later record
+    if (ctx.timestamp() >= ctx.timerService().currentWatermark()) {
+      // ensure every key just register on timer
+      ctx.timerService.registerEventTimeTimer(ctx.timerService.currentWatermark + 1)
+
+      rowState.add(new Tuple2(ctx.timestamp, input))
+    }
+  }
+
+  /**
+    * Called when a timer set fires, sort current records according the timestamp
+    * and emit the output
+    *
+    * @param timestamp The timestamp of the firing timer.
+    * @param ctx       The ctx to register timer or get current time
+    * @param out       The collector for returning result values.
+    */
+  override def onTimer(
+      timestamp: Long,
+      ctx: ProcessFunction[Row, Row]#OnTimerContext,
+      out: Collector[Row]): Unit = {
+
+    var rowList = rowState.get.iterator
+    var sortList = new util.LinkedList[Tuple2[Long, Row]]()
+
+    // sort record according timestamp
+    while (rowList.hasNext) {
+      insertToSortedList(rowList.next, sortList)
+    }
+
+    // emit the output in order
+    if (!sortList.isEmpty) {
+      var lastAccumulator = accumulatorState.value
+      if (lastAccumulator == null) {
+        lastAccumulator = new Row(aggregates.length)
+        for (i <- aggregates.indices) {
+          lastAccumulator.setField(i, aggregates(i).createAccumulator())
+        }
+      }
+
+      var listIter = sortList.listIterator()
+      while (listIter.hasNext) {
+        val curTuple = listIter.next
+        for (i <- 0 until forwardedFieldCount) {
+          output.setField(i, curTuple.f1.getField(i))
+        }
+
+        for (i <- aggregates.indices) {
+          val index = forwardedFieldCount + i
+          val accumulator = lastAccumulator.getField(i).asInstanceOf[Accumulator]
+          aggregates(i).accumulate(accumulator, curTuple.f1.getField(aggFields(i)))
+          output.setField(index, aggregates(i).getValue(accumulator))
+        }
+
+        out.asInstanceOf[TimestampedCollector[Row]].setAbsoluteTimestamp(curTuple.f0)
+        out.collect(output)
+      }
+
+      accumulatorState.update(lastAccumulator)
+      rowState.clear
+      ctx.timerService.registerEventTimeTimer(ctx.timerService.currentWatermark + 1)
```

Comment:
As you suggested, we need to keep elements with timestamp > watermark and register a new timer if the `rowState` is not empty.